### PR TITLE
shell: Add new API for adding subcommands

### PR DESCRIPTION
**Pull request by nordic-krch**
Added macro `SHELL_SUBCMD_SET_CREATE` which creates a set of subcommands.
`SHELL_SUBCMD_ADD` and `SHELL_SUBCMD_COND_ADD` can be used from any file to
add command to the set. This approach allows to have subcommands added
from multiple files.

Solution is using sorted memory section to which subcommands are added. When set is created an empty command is added and because of sorting all commands in the section follows the empty command. Shell is creating an empty entry which is always at the end of the section so that every set of commands is terminated by empty entry of the new command or this global terminator.

Additionally, added dedicated memory section for dynamic subcommands. This allow to change how dynamic commands are distinguished from static one. It is now done based on the address check. Previously it was done by using additional flag.

Added test and extended sample.

### CODE REVIEW

**Comment by nordic-krch** [comment]
Failing CI fails on main branch as well. Created issue: #44544

**Review by jakub-uC — DISMISSED** [review]

File: include/zephyr/shell/shell.h [hunk 1]
```diff
@@ -91,15 +91,12 @@ typedef void (*shell_dynamic_get)(size_t idx,
...
-struct shell_cmd_entry {
-	bool is_dynamic;
-	union union_cmd_entry {
-		/*!< Pointer to function returning dynamic commands.*/
-		shell_dynamic_get dynamic_get;
-
-		/*!< Pointer to array of static commands. */
-		const struct shell_static_entry *entry;
-	} u;
+union shell_cmd_entry {
```
Comment (henrikbrixandersen) [inline include/zephyr/shell/shell.h:94]: We could keep the `struct shell_cmd_entry` having just the union in it for possible future extensions?

**Review by henrikbrixandersen — DISMISSED** [review]

**Comment by carlescufi** [comment]
@nordic-krch please rebase

**Review by jakub-uC — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

File: include/zephyr/shell/shell.h (hunk 1 above)
Comment (henrikbrixandersen) [reply include/zephyr/shell/shell.h:94]: I meant without having to touch each and every user of the union/struct at that time.

**Pull request merged**